src/Balances.sol

```diff
@@ -1,71 +1,75 @@
-import "../lib/openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
 pragma solidity ^0.8.19;
 
-contract Balances is Nonces {
-    event Withdrawal(address indexed account, address target, uint256 value);
-
+contract Balances {
     mapping(address => uint) public balances;
     uint public balanceHeld = 0;
+    address payable[] readyForTransfer;
+    address payable[] scheduledForTransfer;
+    uint public maxTransfersPerTick = 10;
+    uint public minBalanceForTransfer = 1 ether;
+
+    event Transfer(address indexed to, uint256 amount);
+    event TransferFailed(address indexed to, uint256 amount);
 
-    function balanceOf(address account) public view returns (uint) {
-        return balances[account];
+    function balanceOf(address participant) public view returns (uint) {
+        return balances[participant];
     }
 
     function availableBalance() public view returns (uint) {
         return address(this).balance - balanceHeld;
     }
 
-    function releaseBalance(uint amount) internal {
-        balanceHeld -= amount;
-    }
-
-    function reserveBalance(uint amount) internal {
-        // `advanceRound` ensures `amount` doesn't exceed the available balance
+    function increaseParticipantBalance(
+        address payable participant,
+        uint amount
+    ) internal {
+        uint oldBalance = balances[participant];
+        uint newBalance = oldBalance + amount;
+        balances[participant] = newBalance;
         balanceHeld += amount;
+        if (
+            oldBalance <= minBalanceForTransfer &&
+            newBalance > minBalanceForTransfer
+        ) {
+            readyForTransfer.push(participant);
+        }
     }
 
-    function assignBalance(address account, uint amount) internal {
-        balances[account] += amount;
+    function _releaseRewards() internal {
+        scheduledForTransfer = readyForTransfer;
+        delete readyForTransfer;
     }
 
-    function _withdraw(
-        address account,
-        address payable target,
-        uint value
-    ) private {
-        require(balances[account] >= value, "Insufficient balance");
-        balances[account] -= value;
-        balanceHeld -= value;
-        if (balances[account] == 0) {
-            delete balances[account];
+    function transferScheduled() internal {
+        for (
+            uint i = 0;
+            i < scheduledForTransfer.length && i < maxTransfersPerTick;
+            i++
+        ) {
+            address payable participant = scheduledForTransfer[
+                scheduledForTransfer.length - 1
+            ];
+            scheduledForTransfer.pop();
+
+            uint amount = balances[participant];
+            delete balances[participant];
+            balanceHeld -= amount;
+
+            if (participant.send(amount)) {
+                emit Transfer(participant, amount);
+            } else {
+                emit TransferFailed(participant, amount);
```

Comment:
Should we restore `balances[participant]` and `balanceHeld` when the send operation failed?

In the current version, if `participant.send(amount)` fails, the rewards are lost forever.

If we want to avoid infinite loop of sending rewards to a broken address, then perhaps we can assign the amount back to the IE smart contract? Maybe that's how this version already works?

Reply:
The amount never left the IE, so `address(this).balance` still includes it. Now that it has been removed from `balanceHeld`, it's fully available again. So the rewards aren't lost, and yes we need to prevent infinite gas cost by retrying transfers.